### PR TITLE
feat: HD `derivationPath`

### PR DESCRIPTION
**Pull request by davidp94**
Implement #233 

- Added `derivationPath` as an optional parameter to constructor of the clients below.
- Added `derivationPath` as an optional parameter to `setPhrase`

- [x] ethereum  -- example of derivationPath `m/44'/60'/0'/0/5`
- [x] bitcoin -- example of derivationPath `m/84'/0'/0'/0/1`
- [x] bitcoincash -- example of derivationPath `m/44'/145'/0'/0/1`
- [x] litecoin -- example of derivationPath `m/84'/2'/0'/0/1`
- [x] thorchain -- example of derivationPath `44'/931'/0'/0/1`
- [x] binance -- example of derivationPath `2` (this is just the stringified index as the SDK doesn't allow to specify the whole path)

@mikewyszinski notes:

```
    some derivation paths for mainnet/testnet are currently defaulted identical (ex. cosmos), this was to allow the tests to pass as written, to ensure backwards compatibilty with any wallets
    since most of the state has been removed to make managing derivation paths simpler, we added more time so the tests execute more slowly vs. master
    ethereum - master - 80 secs
    ethereum - hd wallet - 180 secs
    there are still parts on the implementations that have the old implementation commented out , if this PR is accepted, i can walk through the code and make sure we delete these
    if the plan is to on-board the top 50-100 coins & support HW wallets in this library, we should consider refactoring the interface to be more "pluggable". This would allow us to build/integrate chains/wallets/explorer/etc as "providers" individually This would make scaling out the lib easier and much easier to maintain.
```

### CODE REVIEW

**Comment by davidp94** [comment]
To reviewers:


Let me know where I should add more comments; documentation; and what other tests cases I should look at; (maybe you can give me your own derivation paths and results from other sources than here!)

For thorchain I am not sure if the addresses path /1 are good; please double check.

**Review by liammcaweeney — APPROVED** [review]
LGTM

**Comment by veado** [comment]
@davidp94 Could you give some more information about your goal with this PR, pls? Do you need to handle different `pathes` or just different `indexes`? Currently `xchain-*` libs have paths hard-coded as defined in #13 to handle different `indexes` (but not `paths`) in the future. Check following note:

> BIP 44 compatibility with other wallets. However the client will only use Index 0, but can in future handle multiple indexes at the same time (eg, like using "sub-accounts" in a single wallet).

^ https://github.com/xchainjs/xchainjs-lib/issues/13

**Comment by davidp94** [comment]
> @davidp94 Could you give some more information about your goal with this PR, pls? Do you need to handle different `paths` or just different `indexes`? Currently `xchain-*` libs have paths hard-coded as defined in #13 to handle different `indexes` (but not `paths`) in the future. Check following note:
> 
> > BIP 44 compatibility with other wallets. However the client will only use Index 0, but can in future handle multiple indexes at the same time (eg, like using "sub-accounts" in a single wallet).
> 
> ^ #13

Goal is to handle different paths (which includes indexes supports); not only indexes.

The fact is that Ledger for ethereum used different paths in legacy version so this approach allows us to be more flexible (cf. https://www.reddit.com/r/ledgerwallet/comments/bpoyoa/which_derivation_path_m44600_or_m4460/ )

Actually #13 might be included for this PR too.

File: packages/xchain-bitcoincash/src/client.ts [hunk 1]
```diff
@@ -289,8 +292,9 @@ class Client implements BitcoinCashClient, XChainClient {
    *
    * @throws {"Invalid address"} Thrown if the given address is an invalid address.
    */
-  getBalance = async (address?: string): Promise<Balance[]> => {
-    return utils.getBalance({ haskoinUrl: this.getHaskoinURL(), address: address || this.getAddress() })
+  getBalance = async (index = 0): Promise<Balance[]> => {
```
Comment (davidp94) [inline packages/xchain-bitcoincash/src/client.ts:295]: Hmm I don't think we should do that for `getBalance`

we should keep `getBalance(address?: string` imo

cc. @veado

**Comment by davidp94** [comment]
Ha yes right; maybe we should still keep getAddress ({address?: string}) ?
if we want to get the balance for another one.

or we would give const secondAddr = getAddress(1)
then getAddress(secondAddr)

File: packages/xchain-cosmos/src/client.ts [hunk 2]
```diff
@@ -48,9 +51,17 @@ class Client implements CosmosClient, XChainClient {
    *
    * @throws {"Invalid phrase"} Thrown if the given phase is invalid.
    */
-  constructor({ network = 'testnet', phrase }: XChainClientParams) {
+  constructor({
+    network = 'testnet',
+    phrase,
+    rootDerivationPaths = {
+      mainnet: `44'/118'/0'/0/`,
+      testnet: `44'/118'/0'/0/`, //TODO: shoudn't this be 44'/118'/0'/1/ ?
```
Comment (davidp94) [inline packages/xchain-cosmos/src/client.ts:59]: Hmm. not sure.

cc, @veado

File: packages/xchain-cosmos/src/client.ts [hunk 3]
```diff
@@ -62,8 +73,8 @@ class Client implements CosmosClient, XChainClient {
    */
   purgeClient(): void {
     this.phrase = ''
-    this.address = ''
-    this.privateKey = null
+    // this.address = ''
```
Comment (davidp94) [inline packages/xchain-cosmos/src/client.ts:76]: Need to remove;

File: packages/xchain-cosmos/src/client.ts [hunk 4]
```diff
@@ -203,27 +219,24 @@ class Client implements CosmosClient, XChainClient {
...
-
   /**
    * Validate the given address.
    *
    * @param {Address} address
    * @returns {boolean} `true` or `false`
    */
   validateAddress = (address: Address): boolean => {
-    return this.sdkClient.checkAddress(address)
+    return this.getCosmosClientSdk(0).checkAddress(address)
```
Comment (davidp94) [inline packages/xchain-cosmos/src/client.ts:239]: This makes performance pretty slow isn't it?

cc @veado

File: packages/xchain-cosmos/src/client.ts [hunk 5]
```diff
@@ -282,9 +294,9 @@ class Client implements CosmosClient, XChainClient {
       this.registerCodecs()
 
       const mainAsset = this.getMainAsset()
-      const txHistory = await this.sdkClient.searchTx({
+      const txHistory = await this.getCosmosClientSdk(0).searchTx({
```
Comment (davidp94) [inline packages/xchain-cosmos/src/client.ts:297]: Again reinstantiating this is heavy; should we have a "anonClient" instantiated used for this kind of function call?

File: packages/xchain-cosmos/src/client.ts [hunk 6]
```diff
@@ -308,7 +320,7 @@ class Client implements CosmosClient, XChainClient {
    */
   getTransactionData = async (txId: string): Promise<Tx> => {
     try {
-      const txResult = await this.sdkClient.txsHashGet(txId)
+      const txResult = await this.getCosmosClientSdk(0).txsHashGet(txId)
```
Comment (davidp94) [inline packages/xchain-cosmos/src/client.ts:323]: Same as previously

File: packages/xchain-ethereum/__tests__/client.mainnet.test.ts [hunk 7]
```diff
@@ -48,67 +64,67 @@ describe('Client Test', () => {
...
+  //       price: {
+  //         rate: 1196.5425814145788,
+  //         diff: 11.71,
+  //         diff7d: 62.3,
+  //         ts: 1609987982,
+  //         marketCapUsd: 136582198332.62915,
+  //         availableSupply: 114147377.999,
+  //         volume24h: 44933107598.39366,
+  //         diff30d: 108.688017487141,
+  //         volDiff1: 7.230942506781318,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.mainnet.test.ts:95]: why is it commented out?

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 8]
```diff
@@ -200,11 +201,9 @@ describe('Client Test', () => {
     expect(balance[1].amount.amount().isEqualTo(baseAmount('96713467036431546', 18).amount())).toBeTruthy()
   })
 
-  it('throws error on bad address', async () => {
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:203]: Need to keep that test. `getBalance('badAddress')`

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 9]
```diff
@@ -529,13 +528,15 @@ describe('Client Test', () => {
     )
 
     let isApproved = await ethClient.isApproved(
+      0,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:531]: Why don't we pass the address here?

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 10]
```diff
@@ -529,13 +528,15 @@ describe('Client Test', () => {
...
       '0x8c2a90d36ec9f745c9b28b588cba5e2a978a1656',
       '0x4f96fe3b7a6cf9725f59d353f723c1bdb64ca6aa',
       baseAmount(100, ETH_DECIMAL),
     )
     expect(isApproved).toEqual(true)
 
     isApproved = await ethClient.isApproved(
+      0,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:539]: same as above

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 11]
```diff
@@ -555,6 +556,7 @@ describe('Client Test', () => {
     mock_all_api(etherscanUrl, ropstenInfuraUrl, ropstenAlchemyUrl, 'eth_estimateGas', '0x5208')
 
     const gasLimit = await ethClient.estimateApprove({
+      walletIndex: 0,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:559]: I think we can put address: string as a parameter as this function is not signing;

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 12]
```diff
@@ -581,6 +583,7 @@ describe('Client Test', () => {
     )
 
     const tx = await ethClient.approve({
+      walletIndex: 0,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:586]: This seems OK;

File: packages/xchain-ethereum/__tests__/client.testnet.test.ts [hunk 13]
```diff
@@ -600,13 +603,19 @@ describe('Client Test', () => {
...
 
-    const gasLimit = await ethClient.estimateCall('0xd15ffaef3112460bf3bcd81087fcbbce394e2ae7', erc20ABI, 'transfer', [
-      '0x8c2a90d36ec9f745c9b28b588cba5e2a978a1656',
-      BigNumber.from(baseAmount('10000000000000', ETH_DECIMAL).amount().toString()),
-      {
-        from: ethClient.getAddress(),
-      },
-    ])
+    const gasLimit = await ethClient.estimateCall(
+      0,
```
Comment (davidp94) [inline packages/xchain-ethereum/__tests__/client.testnet.test.ts:607]: Should be the address

File: packages/xchain-ethereum/src/client.ts [hunk 14]
```diff
@@ -66,8 +67,15 @@ import {
  */
 export interface EthereumClient {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  call<T>(asset: Address, abi: ethers.ContractInterface, func: string, params: Array<any>): Promise<T>
+  call<T>(
+    walletIndex: number,
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:71]: why is that needed?

File: packages/xchain-ethereum/src/client.ts [hunk 15]
```diff
@@ -78,7 +86,7 @@ export interface EthereumClient {
   estimateGasLimit(params: FeesParams): Promise<BigNumber>
   estimateFeesWithGasPricesAndLimits(params: FeesParams): Promise<FeesWithGasPricesAndLimits>
 
-  isApproved(spender: Address, sender: Address, amount: BaseAmount): Promise<boolean>
+  isApproved(walletIndex: number, spender: Address, sender: Address, amount: BaseAmount): Promise<boolean>
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:89]: I don't think we need walletIndex for that call

File: packages/xchain-ethereum/src/client.ts [hunk 16]
```diff
@@ -118,36 +124,22 @@ export default class Client implements XChainClient, EthereumClient {
     ethplorerUrl = 'https://api.ethplorer.io',
     ethplorerApiKey = 'freekey',
     explorerUrl,
-    phrase,
+    phrase = '',
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:127]: again; what happen if phrase = ''?

File: packages/xchain-ethereum/src/client.ts [hunk 17]
```diff
@@ -118,36 +124,22 @@ export default class Client implements XChainClient, EthereumClient {
...
+    rootDerivationPaths = {
+      mainnet: `m/44'/60'/0'/0/`,
+      testnet: `m/44'/60'/0'/0/`, // this is INCORRECT but makes the unit tests pass
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:130]: Why is this incorrect? mainnet and testnet derivation paths are the same for ETH mainnet or testnet. 

https://iancoleman.io/bip39/

File: packages/xchain-ethereum/src/client.ts [hunk 18]
```diff
@@ -187,11 +178,14 @@ export default class Client implements XChainClient, EthereumClient {
...
-    if (!this.address) {
+  getAddress = (index = 0): Address => {
+    if (index < 0) {
+      throw new Error('index must be greater than zero')
+    }
+    if (!this.phrase) {
       throw new Error('Phrase must be provided')
     }
-    return this.address
+    return this.getWallet(index).address.toLowerCase()
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:188]: Any particular reason of removing the checksum casing of the address?

cf. https://coincodex.com/article/2078/ethereum-address-checksum-explained/

File: packages/xchain-ethereum/src/client.ts [hunk 19]
```diff
@@ -570,16 +563,17 @@ export default class Client implements XChainClient, EthereumClient {
    * Thrown if the given contract address is empty.
    */
   estimateCall = async (
-    address: Address,
+    walletIndex = 0,
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:566]: For an estimate call; we don't need to have the wallet seed phrase

We can just do

```
contract.estimateGas[func](...params)
```

And add `from: senderAddress` 

```
ct for a read-only method may include any of:

    overrides.from - the msg.sender (or CALLER) to use during the execution of the code
    overrides.value - the msg.value (or CALLVALUE) to use during the exectuiont of the code
    overrides.gasPrice - the price to pay per gas (theoretically); since there is no transaction, there is not going to be any fee charged, but the EVM still requires a value to report to tx.gasprice (or GASPRICE); most developers will not require this
    overrides.gasLimit - the amount of gas (theoretically) to allow a node to use during the execution of the code; since there is no transaction, there is not going to be any fee charged, but the EVM still processes gas metering so calls like gasleft (or GAS) report meaningful values
    overrides.blockTag - a block tag to simulate the execution at, which can be used for hypothetical historic analysis; note that many backends do not support this, or may require paid plans to access as the node database storage and processing requirements are much higher 
    ```

cf. https://docs.ethers.io/v5/api/contract/contract/

File: packages/xchain-ethereum/src/client.ts [hunk 20]
```diff
@@ -646,10 +650,15 @@ export default class Client implements XChainClient, EthereumClient {
    * @param {BaseAmount} amount The amount of token. By default, it will be unlimited token allowance. (optional)
    * @returns {BigNumber} The estimated gas limit.
    */
-  estimateApprove = async ({ spender, sender, amount }: Omit<ApproveParams, 'feeOptionKey'>): Promise<BigNumber> => {
+  estimateApprove = async ({
+    walletIndex = 0,
```
Comment (davidp94) [inline packages/xchain-ethereum/src/client.ts:654]: estimation does not need that; same as `isApproved` and `estimateCall`

File: packages/xchain-thorchain/src/types/client-types.ts [hunk 21]
```diff
@@ -21,6 +21,7 @@ export type ThorchainClientParams = {
 }
 
 export type DepositParam = {
+  from?: number
```
Comment (davidp94) [inline packages/xchain-thorchain/src/types/client-types.ts:24]: walletIndex would be better

**Review by davidp94 — COMMENTED** [review]
I think there's many area where we have `walletIndex` set up; especially non signing operations.

Need to make naming consistent.

Perforamnce wise; regenerating a wallet for some query calls is not a good thing
